### ISSUE TITLE
mt.__tostring and mt.__serialize called even when opts.metatostring == false

### ISSUE THREAD

**Issue by Rseding91** [pkulchenko/serpent#29]
When opts.metatostring is false mt.__tostring and mt.__serialize are still called on every meta table generating a large amount of extra Lua errors when those don't exist on the meta table.

**Comment by pkulchenko** [comment]
@Rseding91, good catch! It looks like the check needs to be done a bit earlier. Does the following patch works for you?

```diff
diff --git a/src/serpent.lua b/src/serpent.lua
index 4558a1b..4281e5b 100644
--- a/src/serpent.lua
+++ b/src/serpent.lua
@@ -1,4 +1,4 @@
-local n, v = "serpent", "0.30" -- (C) 2012-17 Paul Kulchenko; MIT License
+local n, v = "serpent", "0.31" -- (C) 2012-17 Paul Kulchenko; MIT License
 local c, d = "Paul Kulchenko", "Lua serializer and pretty printer"
 local snum = {[tostring(1/0)]='1/0 --[[math.huge]]',[tostring(-1/0)]='-1/0 --[[-math.huge]]',[tostring(0/0)]='0/0'}
 local badtype = {thread = true, userdata = true, cdata = true}
@@ -51,10 +51,10 @@ local function s(t, opts)
       sref[#sref+1] = spath..space..'='..space..seen[t]
       return tag..'nil'..comment('ref', level) end
     -- protect from those cases where __tostring may fail
-    if type(mt) == 'table' then
+    if type(mt) == 'table' and metatostring ~= false then
       local to, tr = pcall(function() return mt.__tostring(t) end)
       local so, sr = pcall(function() return mt.__serialize(t) end)
-      if (opts.metatostring ~= false and to or so) then -- knows how to serialize itself
+      if (to or so) then -- knows how to serialize itself
         seen[t] = insref or spath
         t = so and sr or tr
         ttype = type(t)
```

BTW, what's the use case when you get errors on those methods?

**Comment by Rseding91** [comment]
Yeah that fixes it. Internally I changed it so it checks metatostring before calling getmetatable(...) as well since the mt variable goes unused other than that if block.

Regarding errors: we compile lua as C++ which means any pcall failure throws an exception internally and is then caught and the lua state returns to what it was doing. I noticed  the VS output window showing thousands of exceptions being thrown during the loading stage where it would use serpent to restore a bunch of script state and narrowed it down to this.
